File: android/app/src/main/java/com/kreativekoala/scribeai/onboarding/OnboardingManager.kt
package com.kreativekoala.scribeai.onboarding

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.kreativekoala.scribeai.data.api.RetrofitClient
import com.kreativekoala.scribeai.data.models.OnboardingPreferencesRequest
import com.kreativekoala.scribeai.utils.AnalyticsService
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * User types for onboarding
 */
enum class UserType(val displayName: String, val value: String) {
    UNDERGRADUATE_STUDENT("Undergraduate Student", "undergraduate_student"),
    HIGH_SCHOOL_STUDENT("High School Student", "high_school_student"),
    MIDDLE_SCHOOL_STUDENT("Middle School Student", "middle_school_student"),
    GRADUATE_STUDENT("Graduate Student", "graduate_student"),
    PROFESSIONAL("Professional", "professional"),
    EDUCATOR("Educator", "educator"),
    OTHER("Other", "other")
}

/**
 * Use cases for onboarding
 */
enum class UseCase(val displayName: String, val value: String, val icon: String) {
    LECTURE_NOTES("Lecture Notes", "lecture_notes", "school"),
    STUDY_MATERIALS("Study Materials", "study_materials", "menu_book"),
    MEETING_NOTES("Meeting Notes", "meeting_notes", "groups"),
    RESEARCH("Research", "research", "search"),
    PERSONAL_LEARNING("Personal Learning", "personal_learning", "psychology"),
    OTHER("Other", "other", "more_horiz")
}

/**
 * Onboarding steps
 */
enum class OnboardingStep(val index: Int) {
    USER_TYPE(0),
    USE_CASE(1),
    FEATURE_UPLOAD(2),
    FEATURE_NOTES(3),
    FEATURE_FLASHCARDS(4),
    FEATURE_QUIZ(5),
    FEATURE_AUDIO(6),
    SOCIAL_PROOF(7),
    COMPARISON(8),
    TRIAL(9),
    NOTIFICATIONS(10);

    val totalSteps: Int get() = values().size

    val progress: Float get() = (index + 1).toFloat() / totalSteps.toFloat()

    fun next(): OnboardingStep? {
        return values().find { it.index == this.index + 1 }
    }

    fun previous(): OnboardingStep? {
        return values().find { it.index == this.index - 1 }
    }
}

/**
 * Manages onboarding state and user preferences
 */
class OnboardingManager(context: Context) {

    companion object {
        private const val TAG = "OnboardingManager"
        private const val PREFS_NAME = "onboarding_prefs"
        private const val KEY_HAS_COMPLETED = "has_completed_onboarding"
        private const val KEY_USER_TYPE = "user_type"
        private const val KEY_USE_CASES = "use_cases"
        private const val KEY_PREFERENCES_SYNCED = "preferences_synced"

        @Volatile
        private var instance: OnboardingManager? = null

        fun getInstance(context: Context): OnboardingManager {
            return instance ?: synchronized(this) {
                instance ?: OnboardingManager(context.applicationContext).also { instance = it }
            }
        }
    }

    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    // State
    private val _currentStep = MutableStateFlow(OnboardingStep.USER_TYPE)
    val currentStep: StateFlow<OnboardingStep> = _currentStep.asStateFlow()

    private val _selectedUserType = MutableStateFlow<UserType?>(null)
    val selectedUserType: StateFlow<UserType?> = _selectedUserType.asStateFlow()

    private val _selectedUseCases = MutableStateFlow<Set<UseCase>>(emptySet())
    val selectedUseCases: StateFlow<Set<UseCase>> = _selectedUseCases.asStateFlow()

    private val _hasCompletedOnboarding = MutableStateFlow(false)
    val hasCompletedOnboarding: StateFlow<Boolean> = _hasCompletedOnboarding.asStateFlow()

    init {
        loadSavedState()
    }

    private fun loadSavedState() {
        _hasCompletedOnboarding.value = prefs.getBoolean(KEY_HAS_COMPLETED, false)

        // Load user type
        prefs.getString(KEY_USER_TYPE, null)?.let { value ->
            _selectedUserType.value = UserType.values().find { it.value == value }
        }

        // Load use cases
        prefs.getStringSet(KEY_USE_CASES, emptySet())?.let { values ->
            _selectedUseCases.value = values.mapNotNull { value ->
                UseCase.values().find { it.value == value }
            }.toSet()
        }
    }

    // MARK: - User Selection

    fun selectUserType(userType: UserType) {
        _selectedUserType.value = userType
    }

    fun toggleUseCase(useCase: UseCase) {
        val current = _selectedUseCases.value.toMutableSet()
        if (current.contains(useCase)) {
            current.remove(useCase)
        } else {
            current.add(useCase)
        }
        _selectedUseCases.value = current
    }

    // MARK: - Navigation

    fun nextStep() {
        val next = _currentStep.value.next()
        if (next != null) {
            _currentStep.value = next
        } else {
            completeOnboarding()
        }
    }

    fun previousStep() {
        _currentStep.value.previous()?.let {
            _currentStep.value = it
        }
    }

    fun goToStep(step: OnboardingStep) {
        _currentStep.value = step
    }

    fun skipOnboarding() {
        completeOnboarding()
    }

    fun skipTrial() {
        // Move to notifications or complete
        if (_currentStep.value == OnboardingStep.TRIAL) {
            nextStep()
        }
    }

    // MARK: - Completion

    private fun completeOnboarding() {
        _hasCompletedOnboarding.value = true

        // Save to preferences
        prefs.edit().apply {
            putBoolean(KEY_HAS_COMPLETED, true)
            _selectedUserType.value?.let { putString(KEY_USER_TYPE, it.value) }
            _selectedUseCases.value.let { useCases ->
                putStringSet(KEY_USE_CASES, useCases.map { it.value }.toSet())
            }
            apply()
        }
    }

    fun markOnboardingComplete() {
        completeOnboarding()
    }

    // MARK: - Backend Sync

    /**
     * Check if preferences need to be synced to backend
     */
    fun needsSync(): Boolean {
        return _hasCompletedOnboarding.value && !prefs.getBoolean(KEY_PREFERENCES_SYNCED, false)
    }

    /**
     * Sync onboarding preferences to backend
     * Call this after user successfully logs in
     */
    fun syncPreferencesToBackend(token: String) {
        if (!needsSync()) {
            Log.d(TAG, "Preferences already synced or onboarding not completed")
            return
        }

        scope.launch {
            try {
                val request = OnboardingPreferencesRequest(
                    userType = _selectedUserType.value?.value,
                    useCases = _selectedUseCases.value.map { it.value }
                )

                val response = RetrofitClient.apiService.saveOnboardingPreferences(
                    "Bearer $token",
                    request
                )

                if (response.isSuccessful && response.body()?.success == true) {
                    prefs.edit().putBoolean(KEY_PREFERENCES_SYNCED, true).apply()
                    Log.d(TAG, "Onboarding preferences synced to backend successfully")

                    // Track analytics
                    AnalyticsService.trackOnboardingCompleted(
                        userType = _selectedUserType.value?.value,
                        useCases = _selectedUseCases.value.map { it.value }
                    )
                } else {
                    Log.e(TAG, "Failed to sync preferences: ${response.errorBody()?.string()}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error syncing preferences to backend", e)
            }
        }
    }

    /**
     * Get saved user type value for analytics/backend
     */
    fun getSavedUserType(): String? {
        return prefs.getString(KEY_USER_TYPE, null)
    }

    /**
     * Get saved use cases for analytics/backend
     */
    fun getSavedUseCases(): List<String> {
        return prefs.getStringSet(KEY_USE_CASES, emptySet())?.toList() ?: emptyList()
    }

    // MARK: - Reset (for testing)

    fun reset() {
        _currentStep.value = OnboardingStep.USER_TYPE
        _selectedUserType.value = null
        _selectedUseCases.value = emptySet()
        _hasCompletedOnboarding.value = false

        prefs.edit().clear().apply()
    }
}

File: ios/scribeai/Onboarding/OnboardingManager.swift
//
//  OnboardingManager.swift
//  scribeai
//
//  Manages onboarding state and user preferences
//

import Foundation
import SwiftUI
import StoreKit

// MARK: - User Types
enum UserType: String, CaseIterable, Codable {
    case undergraduateStudent = "undergraduate_student"
    case highSchoolStudent = "high_school_student"
    case middleSchoolStudent = "middle_school_student"
    case graduateStudent = "graduate_student"
    case professional = "professional"
    case educator = "educator"
    case other = "other"

    var displayName: String {
        switch self {
        case .undergraduateStudent: return "Undergraduate Student"
        case .highSchoolStudent: return "High School Student"
        case .middleSchoolStudent: return "Middle School Student"
        case .graduateStudent: return "Graduate Student"
        case .professional: return "Professional"
        case .educator: return "Educator"
        case .other: return "Other"
        }
    }
}

// MARK: - Use Cases
enum UseCase: String, CaseIterable, Codable {
    case lectureNotes = "lecture_notes"
    case studyMaterials = "study_materials"
    case meetingNotes = "meeting_notes"
    case research = "research"
    case personalLearning = "personal_learning"
    case other = "other"

    var displayName: String {
        switch self {
        case .lectureNotes: return "Lecture Notes"
        case .studyMaterials: return "Study Materials"
        case .meetingNotes: return "Meeting Notes"
        case .research: return "Research"
        case .personalLearning: return "Personal Learning"
        case .other: return "Other"
        }
    }

    var icon: String {
        switch self {
        case .lectureNotes: return "graduationcap"
        case .studyMaterials: return "book"
        case .meetingNotes: return "person.3"
        case .research: return "magnifyingglass"
        case .personalLearning: return "brain"
        case .other: return "ellipsis.circle"
        }
    }
}

// MARK: - Onboarding Step
enum OnboardingStep: Int, CaseIterable {
    case userType = 0
    case useCase = 1
    case featureUpload = 2
    case featureNotes = 3
    case featureFlashcards = 4
    case featureQuiz = 5
    case featureAudio = 6
    case socialProof = 7
    case comparison = 8
    case trial = 9
    case notifications = 10

    var totalSteps: Int { OnboardingStep.allCases.count }

    var progress: Double {
        Double(self.rawValue + 1) / Double(totalSteps)
    }
}

// MARK: - Deletion Reasons
enum DeletionReason: String, CaseIterable, Codable {
    case notUseful = "not_useful"
    case tooExpensive = "too_expensive"
    case betterAlternative = "better_alternative"
    case privacyConcerns = "privacy_concerns"
    case technicalIssues = "technical_issues"
    case other = "other"

    var displayName: String {
        switch self {
        case .notUseful: return "Not useful for my needs"
        case .tooExpensive: return "Too expensive"
        case .betterAlternative: return "Found a better alternative"
        case .privacyConcerns: return "Privacy concerns"
        case .technicalIssues: return "Technical issues"
        case .other: return "Other"
        }
    }
}

// MARK: - Onboarding Manager
@MainActor
class OnboardingManager: ObservableObject {
    static let shared = OnboardingManager()

    // MARK: - Published Properties
    @Published var currentStep: OnboardingStep = .userType
    @Published var selectedUserType: UserType?
    @Published var selectedUseCases: Set<UseCase> = []
    @Published var hasCompletedOnboarding: Bool
    @Published var isLoading = false

    // MARK: - Keys
    private let hasCompletedOnboardingKey = "hasCompletedOnboarding"
    private let userTypeKey = "userType"
    private let useCasesKey = "useCases"
    private let preferencesSyncedKey = "preferencesSynced"

    private init() {
        self.hasCompletedOnboarding = UserDefaults.standard.bool(forKey: hasCompletedOnboardingKey)

        // Load saved preferences
        if let userTypeRaw = UserDefaults.standard.string(forKey: userTypeKey),
           let userType = UserType(rawValue: userTypeRaw) {
            self.selectedUserType = userType
        }

        if let useCasesData = UserDefaults.standard.data(forKey: useCasesKey),
           let useCases = try? JSONDecoder().decode(Set<UseCase>.self, from: useCasesData) {
            self.selectedUseCases = useCases
        }
    }

    // MARK: - Navigation
    func nextStep() {
        guard let nextStep = OnboardingStep(rawValue: currentStep.rawValue + 1) else {
            completeOnboarding()
            return
        }
        withAnimation(.easeInOut(duration: 0.3)) {
            currentStep = nextStep
        }
        trackStepCompleted()
    }

    func previousStep() {
        guard let prevStep = OnboardingStep(rawValue: currentStep.rawValue - 1) else {
            return
        }
        withAnimation(.easeInOut(duration: 0.3)) {
            currentStep = prevStep
        }
    }

    func skipOnboarding() {
        AnalyticsService.shared.track(.onboardingSkipped, properties: [
            "at_step": currentStep.rawValue,
            "step_name": String(describing: currentStep)
        ])
        completeOnboarding()
    }

    func skipTrial(timeSpentSeconds: Int = 0, selectedPlan: String? = nil) {
        AnalyticsService.shared.trackTrialScreenSkipped(
            source: "onboarding",
            timeSpentSeconds: timeSpentSeconds,
            selectedPlan: selectedPlan
        )
        // Move to notifications or complete
        if currentStep == .trial {
            nextStep()
        }
    }

    // MARK: - Completion
    func completeOnboarding() {
        hasCompletedOnboarding = true
        UserDefaults.standard.set(true, forKey: hasCompletedOnboardingKey)

        // Save preferences locally
        if let userType = selectedUserType {
            UserDefaults.standard.set(userType.rawValue, forKey: userTypeKey)
        }

        if let useCasesData = try? JSONEncoder().encode(selectedUseCases) {
            UserDefaults.standard.set(useCasesData, forKey: useCasesKey)
        }

        // Send to backend
        Task {
            await savePreferencesToBackend()
        }

        // Track completion
        AnalyticsService.shared.track(.onboardingCompleted, properties: [
            "user_type": selectedUserType?.rawValue ?? "not_selected",
            "use_cases": selectedUseCases.map { $0.rawValue }
        ])
    }

    // MARK: - Analytics
    private func trackStepCompleted() {
        var properties: [String: Any] = [
            "step": currentStep.rawValue,
            "step_name": String(describing: currentStep)
        ]

        if currentStep == .userType, let userType = selectedUserType {
            properties["selection"] = userType.rawValue
        }

        if currentStep == .useCase {
            properties["selections"] = selectedUseCases.map { $0.rawValue }
        }

        AnalyticsService.shared.track(.onboardingStepCompleted, properties: properties)
    }

    // MARK: - Backend Sync

    /// Check if preferences need to be synced to backend
    var needsSync: Bool {
        hasCompletedOnboarding && !UserDefaults.standard.bool(forKey: preferencesSyncedKey)
    }

    /// Sync preferences to backend - call this after user logs in
    func syncPreferencesToBackendIfNeeded() {
        guard needsSync else {
            print("Onboarding preferences already synced or not completed")
            return
        }

        Task {
            await savePreferencesToBackend()
        }
    }

    private func savePreferencesToBackend() async {
        guard let token = await TokenManager.shared.getValidToken() else {
            print("⚠️ No valid token, cannot sync onboarding preferences")
            return
        }

        let preferences: [String: Any] = [
            "user_type": selectedUserType?.rawValue ?? "",
            "use_cases": selectedUseCases.map { $0.rawValue }
        ]

        do {
            try await APIService.shared.saveOnboardingPreferences(token: token, preferences: preferences)
            UserDefaults.standard.set(true, forKey: preferencesSyncedKey)
            print("✅ Onboarding preferences saved to backend")
        } catch {
            print("⚠️ Failed to save onboarding preferences: \(error)")
        }
    }

    /// Get saved user type for analytics
    func getSavedUserType() -> String? {
        return UserDefaults.standard.string(forKey: userTypeKey)
    }

    /// Get saved use cases for analytics
    func getSavedUseCases() -> [String] {
        if let data = UserDefaults.standard.data(forKey: useCasesKey),
           let useCases = try? JSONDecoder().decode(Set<UseCase>.self, from: data) {
            return useCases.map { $0.rawValue }
        }
        return []
    }

    // MARK: - Reset (for testing)
    func reset() {
        currentStep = .userType
        selectedUserType = nil
        selectedUseCases = []
        hasCompletedOnboarding = false
        UserDefaults.standard.removeObject(forKey: hasCompletedOnboardingKey)
        UserDefaults.standard.removeObject(forKey: userTypeKey)
        UserDefaults.standard.removeObject(forKey: useCasesKey)
        UserDefaults.standard.removeObject(forKey: preferencesSyncedKey)
    }
}

// MARK: - Price Formatting Extension
extension StoreKitManager {
    /// Format price as per-week for better perception
    func weeklyPriceString(for product: Product) -> String? {
        let weeklyPrice: Decimal

        if product.id == SubscriptionProduct.monthly.rawValue {
            // Monthly price / 4.33 weeks
            weeklyPrice = product.price / Decimal(4.33)
        } else if product.id == SubscriptionProduct.yearly.rawValue {
            // Yearly price / 52 weeks
            weeklyPrice = product.price / Decimal(52)
        } else {
            return nil
        }

        return weeklyPrice.formatted(.currency(code: product.priceFormatStyle.currencyCode ?? "USD"))
    }
}
